Index: cpackets.lua
===================================================================
--- cpackets.lua	(revision 61)
+++ cpackets.lua	(working copy)
@@ -1,16 +1,18 @@
 ]]
 [SID_AUTH_CHECK] = { -- 0x51
-	uint32("Client Token"),
-	uint32("EXE Version"),
-	uint32("EXE Hash"),
-	uint32("Number of CD-keys in this packet"),
+	uint32("Client Token", base.HEX),
+	version("EXE Version"),
+	uint32("EXE Hash", base.HEX),
+	uint32{label="Number of CD-keys in this packet", key="cdkeys"},
 	uint32{label="Spawn CD-key", desc=Descs.YesNo},
-	uint32("Key Length"),
-	uint32("CD-key's product value"),
-	uint32("CD-key's public value"),
-	uint32("Unknown"),
-	uint32("[5] Hashed Key Data"),
-	stringz("Exe Information"),
-	stringz("CD-Key owner name"),
+	iterator{label="CD-Key", refkey="cdkeys", repeated={
+		uint32("Key Length"),
+		uint32("CD-key's product value", base.HEX),
+		uint32("CD-key's public value", base.HEX),
+		uint32("Unknown", base.HEX),
+		array{of=uint32, num=5, label="Hashed Key Data"},
+		stringz("Exe Information"),
+		stringz("CD-Key owner name"),
+	}},
 },
 --[[doc
